src/webgpu/shader/validation/decl/var_access_mode.spec.ts

```diff
@@ -0,0 +1,226 @@
+
+    const ok =
+      // The address space must be explicitly specified.
+      t.params.explicitSpace &&
+      // The address space must allow an access mode to be spelled, and the
+      // access mode must be in the list of modes for the address space.
+      t.params.info.spellAccessMode !== 'never' &&
+      (t.params.info.accessModes as readonly string[]).includes(t.params.accessMode);
+
+    t.expectCompileResult(ok, prog);
+  });
+
+g.test('implicit_access_mode')
+  .desc('Validate an implicit access mode on a variable declaration')
+  .params(
+    u =>
+      u
+        .combine('addressSpace', kOrdinaryAddressSpaces)
+        .expand('info', infoExpander)
+        .expand('explicitSpace', explicitSpaceExpander)
+        .combine('explicitMode', [false])
+        .combine('stage', ['compute' as ShaderStage]) // Only need to check compute shaders
+  )
+  .fn(t => {
+    const prog = getShader(t.params);
+
+    // 7.3 var Declarations
+    // "The access mode always has a default value,.."
+    const ok = true;
+
+    t.expectCompileResult(ok, prog);
+  });
+
+g.test('read_access')
+  .desc('A variable can be read from when the access mode permits')
+  .params(
+    u =>
+      u
+        .combine('addressSpace', kOrdinaryAddressSpaces)
+        .expand('info', infoExpander)
+        .expand('explicitSpace', explicitSpaceExpander)
+        .combine('explicitMode', [false, true])
+        .expand('accessMode', accessModeExpander)
+        .combine('stage', ['compute' as ShaderStage]) // Only need to check compute shaders
+  )
+  .fn(t => {
+    const prog = getShader(t.params, 'let copy = x;');
+    const ok = supportsRead(t.params);
+    t.expectCompileResult(ok, prog);
+  });
+
+g.test('write_access')
+  .desc('A variable can be written to when the access mode permits')
+  .params(
+    u =>
+      u
+        .combine('addressSpace', kOrdinaryAddressSpaces)
+        .expand('info', infoExpander)
+        .expand('explicitSpace', explicitSpaceExpander)
+        .combine('explicitMode', [false, true])
+        .expand('accessMode', accessModeExpander)
+        .combine('stage', ['compute' as ShaderStage]) // Only need to check compute shaders
+  )
+  .fn(t => {
+    const prog = getShader(t.params, 'x = 0;');
+    const ok = supportsWrite(t.params);
+    t.expectCompileResult(ok, prog);
+  });
+
+g.test('bad_template_contents')
+  .desc('A variable declaration has explicit access mode with varying other template list contents')
+  .params(u =>
+    u
+      .combine('accessMode', ['read', 'read_write'])
+      .combine('prefix', ['storage,', '', ','])
+      .combine('suffix', [',storage', ',read', ',', ''])
+  )
+  .fn(t => {
+    const prog = `@group(0) @binding(0)
+                  var<${t.params.prefix}${t.params.accessMode}${t.params.suffix}> x: i32;`;
+    const ok = t.params.prefix === 'storage,' && t.params.suffix === '';
+    t.expectCompileResult(ok, prog);
+  });
+
+g.test('bad_template_delim')
```

Comment:
This feels more like a parser test than the rest of the code. Might be better suited in `src/webgpu/shader/validation/parse/var_and_let.spec.ts`